File: src/main/java/gameoflife/algorithm/Cell.java
package gameoflife.algorithm;

import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Cell implements Comparable<Cell> {
    private final String string;
    public final int x, y;

    public Cell(int x, int y) {
        this.x = x;
        this.y = y;
        string = getString(x, y);
    }

    Stream<Cell> getNeighbours() {
        return IntStream.rangeClosed(y - 1, y + 1)
                .mapToObj(this::getNeighboursByRow)
                .flatMap(s -> s);
    }

    private Stream<Cell> getNeighboursByRow(int y) {
        return IntStream.rangeClosed(x - 1, x + 1)
                .mapToObj(x -> new Cell(x, y))
                .filter(this::isNotThis);
    }

    private boolean isNotThis(Cell that) {
        return !equals(that);
    }

    @Override
    public int hashCode() {
        return string.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) {
            return true;
        } else if (obj instanceof Cell) {
            Cell that = (Cell)obj;
            return x == that.x && y == that.y;
        }
        return false;
    }

    @Override
    public String toString() {
        return string;
    }

    @Override
    public int compareTo(Cell that) {
        return toString().compareTo(that.toString());
    }

    static String getString(int x, int y) {
        return x + "|" + y;
    }
}

File: src/main/kotlin/gameoflife/algorithm/Cell.kt
package gameoflife.algorithm

open class Cell(val x: Int, val y: Int) : Comparable<Cell> {
    private val name = toName(x, y)

    val neighbours: List<Cell> get() = (y - 1..y + 1).flatMap { it.neighboursByRow }

    private val Int.neighboursByRow get() = (x - 1..x + 1).filter { !equals(it, this) }.map { Cell(it, this) }

    private fun equals(column: Int, row: Int): Boolean = x == column && y == row

    override fun equals(other: Any?) = other is Cell && equals(other.x, other.y)
    override fun hashCode() = name.hashCode()
    override fun toString() = name
    override fun compareTo(other: Cell) = name.compareTo(other.name)
}

fun toName(x: Int, y: Int) = "$x|$y"